### PR TITLE
Fixes powerfist randomly not working when using mixed gas types

### LINKED ISSUE #56213: Powerfist is "out of gas" despite having a full 1000-2000kpa tank inside.
## Round ID:
This happened on 154387 where it got me killed, I then tested offline with a 16 day old version of the game and reproduced it there too.

## Testmerges:
None of these probably matter because I reproduced it in a 16 day old version of the game.
#55534: '/mob/living/proc/Life(delta_time)' by TemporalOroboros at commit 64722d2fea
#56061: 'Modernizes Admin Only Wizard Event, Department Revolt' by tralezab at commit 7d16921164

## Reproduction:
1) Get a powerfist, set it to any power level
2) Fill a large oxygen tank with an air pump (or a can, or just its default pressure)
3) Apply powerfist to someone with any intent.
4) Based on RNG, you will either hit, or receive the message "The power-fist's piston-ram lets out a weak hiss, it needs more gas!"
5) The powerfist will complain about no gas at 2000kpa, 1400kpa, 900kpa, etc, regardless of power setting.

Even with a airpump filled (approx 2000kpa) tank freshly put in the powerfist, I've had to punch someone 5 times (4 no gas messages) to get an actual powerfist attack to trigger.

Other times its the opposite, i can get 4 real hits in, then the 5'th will fail from no gas, I take the tank out to inspect, and it's still at 1400kpa.

### PR DESCRIPTION
**Pull request by RandomGamer123**
## About The Pull Request
Adds a bit of wiggleroom so QUANTIZE being called on gas_mixture.remove() doesn't randomly and occasionally makes the powerfist think there isn't enough gas. This seems to be the only feasible way of fixing it.

## Why It's Good For The Game
Fixes #56213 

## Changelog
:cl:
fix: Power fist no longer occasionally and randomly says it does not have enough gas when it has enough
/:cl:

### CODE REVIEW

File: code/modules/atmospherics/gasmixtures/gas_mixture.dm [hunk 1]
```diff
@@ -3,9 +3,6 @@ What are the archived variables for?
 Calculations are done using the archived variables with the results merged into the regular variables.
 This prevents race conditions that arise based on the order of tile processing.
 */
-#define MINIMUM_HEAT_CAPACITY 0.0003
-#define MINIMUM_MOLE_COUNT 0.01
-#define MOLAR_ACCURACY  1E-4
 /**
  *I feel the need to document what happens here. Basically this is used
  *catch rounding errors, and make gas go away in small portions.
```
Comment (LemonInTheDark) [inline code/modules/atmospherics/gasmixtures/gas_mixture.dm:8]: You left some documentation behind, bring it with you

File: code/__DEFINES/atmospherics.dm [hunk 2]
```diff
@@ -136,6 +136,13 @@
 /// The total visible states
 #define TOTAL_VISIBLE_STATES (FACTOR_GAS_VISIBLE_MAX * (1 / MOLES_GAS_VISIBLE_STEP))
 
+/// The minimum heat capacity
+#define MINIMUM_HEAT_CAPACITY 0.0003
+/// Minimum mole count of a gas
+#define MINIMUM_MOLE_COUNT 0.01
+/// Molar accuracy to round to
+#define MOLAR_ACCURACY  1E-4
+
```
Comment (LemonInTheDark) [inline code/__DEFINES/atmospherics.dm:145]: Put this up near TOC and friends, I'd classify this as a don't touch if you're not totally sure of what you're doing

File: code/game/objects/items/powerfist.dm [hunk 3]
```diff
@@ -89,7 +89,7 @@
 		target.visible_message("<span class='danger'>[user]'s powerfist lets out a dull thunk as [user.p_they()] punch[user.p_es()] [target.name]!</span>", \
 			"<span class='userdanger'>[user]'s punches you!</span>")
 		return
-	if(gasused.total_moles() < gasperfist * fisto_setting)
+	if(gasused.total_moles() < (gasperfist * fisto_setting - (MOLAR_ACCURACY * 5)))
```
Comment (LemonInTheDark) [inline code/game/objects/items/powerfist.dm:92]: How tested is this? Is this a fix, or a patch?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

File: code/game/objects/items/powerfist.dm (hunk 3 above)
Comment (RandomGamer123) [reply code/game/objects/items/powerfist.dm:92]: I have been unable to reproduce the bug after this even after multiple attempts, and given that it is probably caused by QUANTIZE and floating point rounding if my understanding of it is correct, this should be sufficient.

File: code/__DEFINES/atmospherics.dm [hunk 4]
```diff
@@ -27,6 +27,23 @@
 /// -14C - Temperature used for kitchen cold room, medical freezer, etc.
 #define COLD_ROOM_TEMP 259.15
 
+/// The minimum heat capacity of a gas
+#define MINIMUM_HEAT_CAPACITY 0.0003
+/// Minimum mole count of a gas
+#define MINIMUM_MOLE_COUNT 0.01
+/// Molar accuracy to round to
+#define MOLAR_ACCURACY  1E-4
+/**
```
Comment (Mothblocks) [inline code/__DEFINES/atmospherics.dm:36]: This doc-comment is going to be applied to MOLES_CELLSTANDARD. Is that intentional?

File: code/game/objects/items/powerfist.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/game/objects/items/powerfist.dm:92]: Could you figure out a cleaner way of checking to see if  something is within a rounding error from QUANTIZE, and turn it into a define so I can make other people use it?

File: code/__DEFINES/atmospherics.dm (hunk 4 above)
Comment (RandomGamer123) [reply code/__DEFINES/atmospherics.dm:36]: I thought that multiline comments like that weren't documented, i'll move it up over this segment

File: code/game/objects/items/powerfist.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/game/objects/items/powerfist.dm:92]: I can give it a shot if you're having a painful time of it

**Comment by Mothblocks** [comment]
Undraft if ready.

File: code/game/objects/items/powerfist.dm (hunk 3 above)
Comment (RandomGamer123) [reply code/game/objects/items/powerfist.dm]: This works for you?

**Comment by RandomGamer123** [comment]
@LemonInTheDark Could you see if this works or otherwise ill just revert the last change

File: code/__HELPERS/cmp.dm [hunk 5]
```diff
@@ -1,3 +1,12 @@
+/proc/cmp_epsilon_less_than(a,b,epsilon = MOLAR_ACCURACY)
+	return (a < (b + epsilon))
+
+/proc/cmp_epsilon_greater_than(a,b,epsilon = MOLAR_ACCURACY)
+	return ((a + epsilon) > b)
+
+/proc/cmp_epsilon_equals(a,b,epsilon = MOLAR_ACCURACY)
+	return (((a + epsilon) > b) && ((a - epsilon) < b))
+
```
Comment (LemonInTheDark) [inline code/__HELPERS/cmp.dm:9]: These procs are typically used for TimSort, our sorting proc. You're better off turning them into defines. The spirit is right tho

File: code/game/objects/items/powerfist.dm [hunk 6]
```diff
@@ -89,7 +89,7 @@
 		target.visible_message("<span class='danger'>[user]'s powerfist lets out a dull thunk as [user.p_they()] punch[user.p_es()] [target.name]!</span>", \
 			"<span class='userdanger'>[user]'s punches you!</span>")
 		return
-	if(gasused.total_moles() < gasperfist * fisto_setting)
+	if(cmp_epsilon_less_than(gasused.total_moles(), gasperfist * fisto_setting, (MOLAR_ACCURACY * 5)))
```
Comment (LemonInTheDark) [inline code/game/objects/items/powerfist.dm:92]: why the * 5? each of these types of comparison will need it, right? so why not make it part of a proc or define so people don't need to dupe code?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Sorry bout the delay, you don't need to revert this, it's better then it was before, but I think it still needs a slight bit more polish

File: code/game/objects/items/powerfist.dm (hunk 6 above)
Comment (RandomGamer123) [reply code/game/objects/items/powerfist.dm:92]: To be more precise, the 5 is an arbitrary selection. In reality QUANTIZE is being called on every single gas type, so the theoretical largest difference between pre-quantize and post-quantize gas mix is (gas types in mixture - 2)* (MOLAR_ACCURACY) plus a bit more leeway for floating point messiness, however in most in game circumstances and in most gas mixes 5 is more than enough, which is why I chose it

**Comment by Rohesie** [comment]
Still have issues with this, @LemonInTheDark ?

**Comment by RandomGamer123** [comment]
> Still have issues with this, @LemonInTheDark ?

Should probably draft it, not ready yet actually on my end, will complete one day

**Comment by RandomGamer123** [comment]
> > Still have issues with this, @LemonInTheDark ?
> 
> Should probably draft it, not ready yet actually on my end, will complete one day

Well actually on second thought I can't really make equals a define to begin with so... I'll see what Lemon says

**Comment by Rohesie** [comment]
> 
> 
> > Still have issues with this, @LemonInTheDark ?
> 
> Should probably draft it, not ready yet actually on my end, will complete one day

You can draft your own PRs, can't you?

**Comment by RandomGamer123** [comment]
> > > Still have issues with this, @LemonInTheDark ?
> > 
> > 
> > Should probably draft it, not ready yet actually on my end, will complete one day
> 
> You can draft your own PRs, can't you?

Well my second thought was to wait for Lemon's response to see if this can be merged as-is or converting it to a define since one actually can't be converted well.

**Comment by LemonInTheDark** [comment]
The define bit can be dropped, I guess my actual problem is it's a bit too general.

I'd like a proc/define I can just throw at someone who's comparing mole counts, and just have it work with no extra inputs.

Could you make something that's bespoke for comparing mole counts, rather then a general rounding compare thing? You'll want to move it out of the cmps file, since that's used for our sorting algorithm, which  this I hope should never need to be used with

File: code/__HELPERS/atmospherics.dm [hunk 7]
```diff
@@ -0,0 +1,8 @@
+/proc/cmp_epsilon_less_than(a,b,epsilon = MAXIMUM_ERROR_GAS_REMOVAL)
+	return (a < (b + epsilon))
+
+/proc/cmp_epsilon_greater_than(a,b,epsilon = MAXIMUM_ERROR_GAS_REMOVAL)
+	return ((a + epsilon) > b)
+
+/proc/cmp_epsilon_equals(a,b,epsilon = MAXIMUM_ERROR_GAS_REMOVAL)
+	return (((a + epsilon) > b) && ((a - epsilon) < b))
```
Comment (LemonInTheDark) [inline code/__HELPERS/atmospherics.dm:8]: Call it like, molar compare or something, and remove the epsilon arg

Comment (RandomGamer123) [reply code/__HELPERS/atmospherics.dm]: MAXIMUM_ERROR_GAS_REMOVAL is like 2.2e-3 which well while good for most purposes is way overkill for cases involving a single gas type (eg. perhaps filter / breath code) so I would personally consider keeping it, besides, if you don't specify an epsilon it'll default to that anyways, at the cost of well presumably slight performance differences

Comment (LemonInTheDark) [reply code/__HELPERS/atmospherics.dm]: alright that's good enough for me

File: code/game/objects/items/powerfist.dm [hunk 8]
```diff
@@ -89,7 +89,7 @@
 		target.visible_message(span_danger("[user]'s powerfist lets out a dull thunk as [user.p_they()] punch[user.p_es()] [target.name]!"), \
 			span_userdanger("[user]'s punches you!"))
 		return
-	if(gasused.total_moles() < gasperfist * fisto_setting)
+	if(!cmp_epsilon_equals(gasused.total_moles(), gasperfist * fisto_setting))
```
Comment (LemonInTheDark) [inline code/game/objects/items/powerfist.dm:92]: make this the right proc and we're ready to roll

**Pull request merged**